### PR TITLE
Backend Test Suite Setup and Example

### LINKED ISSUE #143: Backend Test Suite Setup
**User Story:**

As a contributor, I want to have some example test cases so I know what type of tests to write in the future.

**Acceptance Criteria:**

A test suite setup, with examples to refer to when adding new tests.

**Additional Context:**

### PR DESCRIPTION
**Pull request by jacoballen4534**
## Link to Issue Number
closes #143 

## Description
Setup the backend test suite and added an example for others to base off of.
Added to the README to outline the steps to add new tests.

## Checklist
- [x] All feature request A/C have been met OR the bug has been fixed
- [x] I ran all necessary tests and they pass
- [x] I rebased upstream/master onto my working branch before opening this PR
- [x] I have named my PR something sensible
- [x] I have included the issue number above
- [x] I have assigned at least two people to review my changes

## Other Notes
No notes.

### CODE REVIEW

File: backend/tests/tests.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import config from '../config/config';
+
+if (!config.mock) {
+  throw '\n\n   Tests should only be run on the mock database. Please set the mock flag in the config to true.\n \
+  If you know what you are doing and wish to run this on a real database, remove this check.\n\n';
+}
+
+// List all tests to be run.
+require('./reservation/getAllForRestaurant'); // Using IDs 1-5
```
Comment (itsercleung) [inline backend/tests/tests.js]: Nitpick: Try not to use inline comments in the future (can be hard to read sometimes)

File: backend/tests/tests.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import config from '../config/config';
```
Comment (itsercleung) [inline backend/tests/tests.js]: Nitpick: The config file could be more concise and not the same name as directory

File: backend/package.json [hunk 3]
```diff
@@ -9,8 +9,8 @@
     "lintfix": "eslint --fix .",
     "start": "babel-node server",
     "mockTest": "bash test.sh",
-    "dockerTest": "mocha tests --recursive --require babel-polyfill --require @babel/register",
-    "hostedTest": "mocha tests --recursive --require babel-polyfill --require @babel/register"
+    "dockerTest": "mocha tests/tests --require babel-polyfill --require @babel/register",
```
Comment (itsercleung) [inline backend/package.json]: Still a little unsure about having two scripts being the same with different names  (`hostedTest`, `dockerTest`)

File: backend/tests/tests.js (hunk 1 above)
Comment (jacoballen4534) [reply backend/tests/tests.js]: I was thinking, as there will be multiple lines all importing a file and declaring an id, having the id comment above might get a bit confusing as to what test it belongs to.

File: backend/package.json (hunk 3 above)
Comment (jacoballen4534) [reply backend/package.json]: I will just remove the hosted option.

**Comment by jacoballen4534** [comment]
> Great work @jacoballen4534. I think it would be very helpful if we have some block comments describing what is the purpose of the tests so that readers can easily understand it. Otherwise looks good.

The section at the top of each test is aimed to briefly explain what the test is doing.
```javascript
  it('1. should ... Explain the aim of the test', async function() {
```
Would you want another block comment with more of an explanation?

File: backend/tests/tests.js (hunk 2 above)
Comment (jacoballen4534) [reply backend/tests/tests.js]: I named the config file index.js as this allows all imports to be simplified to just /config

**Comment by nidhineshnand** [comment]
@jacoballen4534 Yeah, something that describes why we are testing the feature for example "Checks that the API is able to handle calls when there is no reservation".

**Comment by jacoballen4534** [comment]
> @jacoballen4534 Yeah, something that describes why we are testing the feature for example "Checks that the API is able to handle calls when there is no reservation".

Is that not the same as what I currently have.?
"it should return an empty array when there are no reservations"

**Comment by jacoballen4534** [comment]
> Great work @jacoballen4534. I think it would be very helpful if we have some block comments describing what is the purpose of the tests so that readers can easily understand it. Otherwise looks good.

The section at the top of each test is aimed to briefly explain what the test is doing.
```javascript
  it('1. should ... Explain the aim of the test', async function() {
```
Would you want another block comment with more of an explanation?

**Comment by nidhineshnand** [comment]
Currently, it explains what the expected behaviour not the reason for testing that feature.

**Comment by jacoballen4534** [comment]
> Currently, it explains what the expected behaviour not the reason for testing that feature.

Updated.

File: backend/README.md [hunk 4]
```diff
@@ -127,7 +127,7 @@ $ npm i
 
 Note.
 
-- Before running any tests, change the mock flag in `config.js` to true.
+- Before running any tests, change the mock flag the config to true.
```
Comment (gracecocks) [inline backend/README.md]: change the mock flag **in the** config to true?

File: backend/tests/reservation/getAllForRestaurant.js [hunk 5]
```diff
@@ -0,0 +1,264 @@
+import chai, { assert } from 'chai';
+import chaiHttp from 'chai-http';
+import config from '../../config';
+import connection from '../../database';
+
+chai.use(chaiHttp);
+// Testing /reservation/?restaurantID GET endpoint. Using IDs 1-5
+
+before(async () => {
+  // Setup the databse before any tests in this file run.
```
Comment (gracecocks) [inline backend/tests/reservation/getAllForRestaurant.js]: database

File: backend/tests/reservation/getAllForRestaurant.js [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+  const errors = [];
+  await connection
+    .asyncQuery('INSERT INTO restaurant_db.RESTAURANT VALUES (1, "Example resteraunt");')
```
Comment (gracecocks) [inline backend/tests/reservation/getAllForRestaurant.js]: nit pick, but **example restaurant**? or is it supposed to be spelt different as an example of a named restaurant? unsure

**Pull request merged**